README.md
# Okanjo Application Framework

[![Node.js CI](https://github.com/Okanjo/okanjo-app/actions/workflows/node.js.yml/badge.svg)](https://github.com/Okanjo/okanjo-app/actions/workflows/node.js.yml) [![Coverage Status](https://coveralls.io/repos/github/Okanjo/okanjo-app/badge.svg?branch=master)](https://coveralls.io/github/Okanjo/okanjo-app?branch=master)

This module framework helps make creating scalable applications quick and simple, focusing on configuration and flexibility over boilerplate.

The core application module handles:
* Multi-environment, extendable application configuration
* Error reporting (optionally, via [Sentry.io](https://sentry.io))
* Connection and start up initialization handling
* Utility functions 

The principle here is that an application is built upon okanjo-app, leveraging other okanjo modules to interface with 
various services. 

## Installing

Add to your project like so: 

```sh
npm install okanjo-app
```

Requires peer dependencies:

```sh
npm install @hapi/boom raven
```

## Example usage

A very simple implementation could look like this:

```js
const OkanjoApp = require('okanjo-app');
const config = {
    yourThing: {
        yourKey: 'yourValue'
    },
    reportToSentry: false,
    // ravenReportUri: 'http://your-app@app.getsentry.com/number',
    // ravenOptions: {...}
};

const app = new OkanjoApp(config);

// Additional initialization, plugin registering, etc

// Go!
await app.connectToServices();

```

We generally recommend creating a `config.js` file, that exports a configuration object.

```text
my_app/
+- node_modules/
+- routes/
+- config.js
+- index.js
+- package.json
```

config.js:
```js
const Path = require('path');
module.exports = {
    webServer: {
        port: 3000,
        routePath: Path.join(__dirname, 'routes')
    },
    
    reportToSentry: false,
    // ravenReportUri: 'https://your-reporting-uri',
    // ravenOptions: {...}
    
    production: {
        webServer: {
            port: 80
        },
        reportToSentry: true
    }
};
```

index.js:
```js
const OkanjoApp = require('okanjo-app');
const app = new OkanjoApp(config);

app.connectToServices().then(() => {
    // Everything connected, ready to do your thing!
});
```

Then you can start the app like so:
* Default environment (e.g. locally): `node .`
* Production environment: `env=production node .`

## Properties

* `app.config` – The effective configuration of the application.
* `app.currentEnvironment` – The name of the current application environment. (e.g. `default`)
* `app.gracefulShutdown` – Flag your application may use to indicate the app should be shutting down.
* `app.ravenClient` – [Raven](https://docs.sentry.io/clients/node/) client instance
* `app.ready` – Boolean indicating whether `app.connectToServices` was called and completed.
* `app.reportingContext` – Object whose key-value pairs are copied to error reports.

## Methods

### Constructor

`const app = new OkanjoApp(config)`
* `config`: Object with key-values or nested environment configurations

### `await app.connectToServices([callback])`
Starts initialization for any registered services and fires the callback once all are completed.
* `callback(err)`: Function that is fired when application modules finish initializing.
  * `err` – if a service connector returns an error, it'll stop and get caught here
* Returns a `Promise`, so you can await it

### `app.updateReportingStatus(enabled)`
This method is internally called when constructing. You may use this method to enable reporting if initially disabled.

* `enabled`: Boolean flag to turn reporting on (`true`) or off (`false`)

### `app.setReportingContext(context)`
Attaches the given keys within `context` to the reporting context, which is attached to any future error reports. 
This is useful for setting global information on error reports, such as environment, application, current server, etc.
* `context`: Object with key-value pairs. 

### `await app.report(...)`
Used to report application errors or bad things to stderr and if configured, Sentry.

This method will take any number of parameters. 
The last `Error` object encountered will be selected as the error being reported.
If no `Error` object is provided, one will be generated using the first `string` argument provided.

For example:
```js
myService.connect((err) => {
    if (err) {
        app.report('Failed to connect to my service!', err, { additional: 'context'});
    }
    // ...
});
```

Other example usages:
* `await app.report(err)` - just report an error with no context
* `await app.report('what happened', err)` - report what happened with the original error
* `await app.report('what happened', err, { arg, context, whatever })` - report what happened and provide additional data, detailing what the state was at the time of the error

### `app.dump(...)`
Takes any number of arguments, and for each, prints each to stderr with colors and max depth of 5. Useful for quickly debugging or reporting information about a complex object.

Example usages:
* `app.dump(myObject)`
* `app.dump(myThing, myOtherThing, 42)`


### `app.log(...)`
A wrapper around console.log that writes to stderr and can be conditionally disabled by setting environment variable `SILENCE_REPORTS=1`.
Useful for making various logging tasks optional.

Example usages:
* `app.log('hey look at this thing, thing)`;

### `app.copy(destination, source)`
### `OkanjoApp.copy(destination, source)`
Deep copies key-values from `source` to `destination`. Traverse arrays and objects recursively.
* `source` – Source object to copy from
* `destination` – Destination object to copy into. If not truthy, it will be made into an object or array. 

For example:
```js
const source = {
    a: 1,
    b: {
        hi: 'there'
    },
    e: null
};

const destination = {
    c: true,
    b: { there: 'hi' },
    d: null
};

app.copy(destination, source);

/* destination now looks like:
{
    a: 1,
    b: { hi: 'there', there: 'hi' },
    c: true,
    d: null,
    e: null
}
 */

```

### `app.flattenData(input, options)`
### `OkanjoApp.flattenData(input, options)`
Useful for flattening a nested object into a single level object.
* `input` – The object to flatten
* `options` – Optional. Flags that can affect the output.
** `options.dateToIso` – If truthy, then date objects will be flattened into an ISO string.

MongoDB ObjectId's will be converted to hex strings.

For example:
```js
app.flattenData({ a: { b: 1, c: { d: [2,3,4] }}});

/* returns:
{
    a_b: 1,
    a_c_d_0: 2,
    a_c_d_1: 3,
    a_c_d_2: 4 
}
 */

```
 
### `app.response`
OkanjoApp extends [Hapi Boom](https://github.com/hapijs/boom). All boom methods are available as possible responses. 
Since Boom is great for error response building, OkanjoApp extends it to add consistent 'positive' responses.

For example:
```js
app.response.badRequest('You goofed');
/* Reply shows:
{
    "statusCode": 400,
    "error": "Bad Request",
    "message": "You goofed"
}
 */
```
 
### `app.response.ok(data, [h])`
Creates a new 200/ok response with the given data payload.
* `data` – value to output as data in the response.
* `h` – Optional, response toolkit. Internally does: h.response(data).code(200)

For example:
```js
app.response.ok({ hello: 'world' });
/* Reply shows:
{
    "statusCode": 200,
    "error": null,
    "data": { "hello": "world" }
}
 */
```

### `app.response.created(data, [h])`
Creates a new 201/created response with the given data payload.
* `data` – value to output as data in the response.
* `h` – Optional, response toolkit. Internally does: h.response(data).code(201)

For example:
```js
app.response.created({ hello: 'world' });
/* Reply shows:
{
    "statusCode": 201,
    "error": null,
    "data": { "hello": "world" }
}
 */
```

### `app.response.formatForResponse(obj, closure)`
Useful helper function for converting internal data objects into ones suitable for api output.
* `obj` – Object or array of objects to format
* `closure(obj)` – Function which is expected to format and return the a single object

Note: This method automatically attaches `obj.created` and `obj.updated` is present in `obj`.  

For example:
```js
function formatter(mixed) {
    return app.response.formatForResponse(mixed, (obj) => {
        return {
            id: "thing_" + obj._id,
            name: obj.name || "Untitled",
            // ...
        };
    });
}

formatter({
    _id: "1",
    name: "thing 1",
    secret: "you should not see me",
    created: new Date(),
    updated: null
})
/* 
{ id: 'thing_1',
  name: 'thing 1',
  created: 2017-11-07T18:26:09.022Z,
  updated: null }
*/


formatter({
    _id: "2",
    name: null,
    secret: "you should not see me"
})
/* 
{ id: 'thing_2', name: 'Untitled' }
*/
``` 

## Events

### `app.once('ready', callback)`
Fires `callback` when the app has finished connecting to all services and is ready.

### `app.once('error', callback)`
Fires `callback(err)` when a service connector fails on `connectToServices()`


## Extending and Contributing 

Our goal is quality-driven development. Please ensure that 100% of the code is covered with testing.

Before contributing pull requests, please ensure that changes are covered with unit tests, and that all are passing. 

### Testing

To run unit tests and code coverage:
```sh
npm run report
```

This will perform:
* Unit tests
* Code coverage report
* Code linting

Sometimes, that's overkill to quickly test a quick change. To run just the unit tests:
 
```sh
npm test
```

or if you have mocha installed globally, you may run `mocha test` instead.


CHANGELOG.md
# Changelog

## v3.2.0
Node v16 Upgrade!

Notable Changes:
- Using callback functions in service connectors is deprecated. Return a Promise or AsyncFunction instead.

- OkanjoApp.js
    - Removed Async dependency
    - Added config option `ravenOptions` to pass directly into the raven client, which may override defaults
    - connectToServices now uses Promise.all instead of the Async parallel
    - Various refactoring
- package.json:
    - Bumped to 3.2.0
    - Upgraded all dependencies to latest
    - Removed async lib
    - Moved eslint config inside

## v2.0.0
 * Updated for async/await and Node v8
 * Updated dependencies to latest
 * Removed OkanjoBoom and replaced with with latest Boom with .ok and .created hacks
 * `connectToServices` returns a promise so it can be awaited, `callback` is optional
 * `connectToServices` now throws if a service function errors, also emits `error` event
 * Service connectors can be async functions (handled by Async.js)
 * Removed `ifOk` helper, since it's irrelevant in Hapi 17
 * `report(...)` is now an async function which returns after the Sentry call completes. Will not throw an error.
 * Renamed `app.inspect` to `app.dump` to avoid Node 10 deprecation warnings. Custom symbol isn't worth the hassle.

## v1.0.2
 * Fixed triggering custom inspection on app instances (e.g. util.inspect firing app.inspect by mistake)
 
## v1.0.1
 * Corrected module name
 * Swapped out jshint for eslint for consistency with the other apps
 * Updated README for minor corrections

## v1.0.0
 * Initial public release, after 26 internal releases!


test/boom-test.js
const should = require('should');
const Cluster = require("cluster");
if (!Cluster.isMaster) return;

describe('Our Boom', function() {

    const OkanjoApp = require('../OkanjoApp');

    class Response {

        constructor(data) {
            this.output = data;
            this.statusCode = null;
        }

        code(statusCode) {
            this.statusCode = statusCode;
            return this;
        }
    }

    const FakeResponseToolkit = {
        response: (data) => {
            return new Response(data);
        }
    };

    it('can make a success response', function() {
        const res = OkanjoApp.response.ok({
            test: "string",
            collection: [1, 2, 3]
        });

        res.should.be.an.Object();
        res.statusCode.should.be.equal(200);
        should(res.error).be.null();
        res.data.should.be.an.Object();
        res.data.test.should.be.a.String();
        res.data.collection.should.be.an.Array();
        res.data.collection.length.should.be.equal(3);
    });

    it('can make a success response with the response toolkit', function() {
        const res = OkanjoApp.response.ok({
            test: "string",
            collection: [1, 2, 3]
        }, FakeResponseToolkit);

        res.should.be.an.Object();
        res.statusCode.should.be.equal(200);
        res.output.statusCode.should.be.equal(200);
        should(res.output.error).be.null();
        res.output.data.should.be.an.Object();
        res.output.data.test.should.be.a.String();
        res.output.data.collection.should.be.an.Array();
        res.output.data.collection.length.should.be.equal(3);
    });

    it('can make a created response', function() {
        const res = OkanjoApp.response.created({
            test: "string",
            collection: [1, 2, 3]
        });

        res.should.be.an.Object();
        res.statusCode.should.be.equal(201);
        should(res.error).be.null();
        res.data.should.be.an.Object();
        res.data.test.should.be.a.String();
        res.data.collection.should.be.an.Array();
        res.data.collection.length.should.be.equal(3);
    });

    it('can make a created response with the response toolkit', function() {
        const res = OkanjoApp.response.created({
            test: "string",
            collection: [1, 2, 3]
        }, FakeResponseToolkit);

        res.should.be.an.Object();
        res.statusCode.should.be.equal(201);
        res.output.statusCode.should.be.equal(201);
        should(res.output.error).be.null();
        res.output.data.should.be.an.Object();
        res.output.data.test.should.be.a.String();
        res.output.data.collection.should.be.an.Array();
        res.output.data.collection.length.should.be.equal(3);
    });

    it('can format an object using a closure', function() {
        const original = {
            id: "12345",
            name: "wayne",
            status: "sitting next to kevin",
            pants: true,
            created: new Date(),
            updated: new Date()
        };

        const data = OkanjoApp.response.formatForResponse(original, function (obj) {
            return {
                id: "prefix_" + obj.id,
                name: obj.name,
                status: obj.status,
                injected: true
            };
        });

        data.should.be.an.Object();
        data.should.not.be.an.Array();

        // these fields shoulda copied
        data.id.should.equal("prefix_"+original.id);
        data.name.should.be.a.String();
        data.status.should.be.a.String();

        // this field shoulda been stripped
        should(data.pants).not.be.ok();

        // created and updated should automatically copy over
        data.created.should.be.instanceof(Date);
        data.updated.should.be.instanceof(Date);

        // this property should now exist
        data.injected.should.be.equal(true);
    });

    it('can format an array of objects using a closure', function() {

        const original = [
            {
                id: "12345",
                name: "wayne",
                status: "sitting next to kevin",
                pants: true,
                created: new Date()
            },
            {
                id: "123456",
                name: "matt",
                status: "sitting next to wayne",
                pants: true
            },
            {
                id: "123457",
                name: "dave",
                status: "sitting behind to matt",
                pants: false,
                created: new Date()
            }
        ];

        const data = OkanjoApp.response.formatForResponse(original, function (obj) {
            return {
                id: "prefix_" + obj.id,
                name: obj.name,
                status: obj.status,
                created: obj.created,
                injected: true
            };
        });


        data.should.be.an.Array();
        data.forEach(function(data, idx) {
            data.id.should.equal("prefix_"+original[idx].id);
            data.name.should.be.a.String();
            data.status.should.be.a.String();
            should(data.pants).not.be.ok();
            data.injected.should.be.equal(true);
        });

    });

    it('will format empty values as null', function() {
        let data = OkanjoApp.response.formatForResponse(null, function () {
            return "nope";
        });
        should(data).be.exactly(null);

        data = OkanjoApp.response.formatForResponse(undefined, function() {
            return "nope";
        });
        should(data).be.exactly(null);

        data = OkanjoApp.response.formatForResponse("", function() {
            return "nope";
        });
        should(data).be.exactly(null);

        data = OkanjoApp.response.formatForResponse(0, function() {
            return "nope";
        });
        should(data).be.exactly(null);
    });

    it('is accessible in an OkanjoApp instance', function() {
        const app = new OkanjoApp({});

        app.response.should.be.an.Object();

        const res = app.response.ok({
            test: "string",
            collection: [1, 2, 3]
        });

        res.should.be.an.Object();
        res.statusCode.should.be.equal(200);
        should(res.error).be.null();
        res.data.should.be.an.Object();
        res.data.test.should.be.a.String();
        res.data.collection.should.be.an.Array();
        res.data.collection.length.should.be.equal(3);
    });

});

test/app-worker-test.js
const should = require('should');
const Cluster = require('cluster');
const ravenUnitTestAppUri = process.env.SENTRY_URI;

const sharedConfig = {
    ravenReportUri: ravenUnitTestAppUri,
    reportToSentry: true,

    connectionString: "default!",
    otherSetting: "yup",

    service: {
        url: "http://overyet.com"
    },

    dev: {
        connectionString: "dev!",
        devOnlySetting: "yuppers",

        service: {
            url: "http://is.winter.overyet.com"
        }
    }
};

if (Cluster.isMaster) return;

describe('Worker', () => {

    const OkanjoApp = require('../OkanjoApp');

    it('process should have env', () => {
        should(process.env.env).be.a.String().and.not.empty();
    });

    it('instantiates ok', done => {

        const app = new OkanjoApp(sharedConfig);
        app.should.be.an.Object();

        function finish() {

            done();

            process.nextTick(() => {
                process.exit(0);
            });
        }

        if (process.env.mode === "basic") {
            app.currentEnvironment.should.be.equal('default');
            app.reportingContext.environment.should.be.equal('default');
            finish();
        } else if (process.env.mode === "dev") {
            // Check it
            app.currentEnvironment.should.be.equal('dev');
            app.reportingContext.environment.should.be.equal('dev');

            // Check that the config applied correctly
            app.config.connectionString.should.be.equal('dev!');
            app.config.otherSetting.should.be.equal('yup');
            app.config.devOnlySetting.should.be.equal('yuppers');
            app.config.service.url.should.be.equal('http://is.winter.overyet.com');
            finish();
        } else if (process.env.mode === "explode") {
            app._reportUncaughtException(new Error('Fall down go boom with exit'));
        } else {
            throw new Error('Unknown test mode!');
        }

    });

});



OkanjoApp.js
"use strict";

const Util = require('util');
const Raven = require('raven');
const Cluster = require('cluster');
const EventEmitter = require('events').EventEmitter;

/**
 * Common application helpers and configurations
 * @extends EventEmitter
 * @constructor
 */
class OkanjoApp extends EventEmitter {

    //region Setup & Configuration

    /**
     * Constructor
     * @param config – Application configuration
     */
    constructor(config) {
        super();

        // When the app scheduled to quit, this will be true
        this.gracefulShutdown = false;

        this.currentEnvironment = "default";
        this.config = config;

        // Copy static shortcuts
        this.response = OkanjoApp.response;
        this.copy = OkanjoApp.copy;
        this.flattenData = OkanjoApp.flattenData;

        // If we have an environment set in the process env vars, apply environment configuration overrides
        if (process.env.env && process.env.env !== "default") {
            this._applyEnvConfig(process.env.env, this.config);
        }

        // Set Sentry reporting context information
        this._reportToSentry = false;
        this.reportingContext = {
            environment: this.currentEnvironment,
            worker_type: process.env.worker_type || 'master'
        };

        // Create the reporter
        const ravenOptions = this.config.ravenOptions || {};
        this.ravenClient = new Raven.Client(this.config.ravenReportUri, {
            // release: ...
            tags: {
                worker_type: process.env.worker_type || 'master',
            },
            // extra: { key: ... }
            environment: this.currentEnvironment,
            ...ravenOptions
        });
        this._captureException = Util.promisify(this.ravenClient.captureException.bind(this.ravenClient));

        // Start rockin' reports
        this.updateReportingStatus(this.config.reportToSentry || false);

        //
        // Connection initialization
        //

        // Status var that determines whether the app is in the process of connecting to requisite services
        this._connecting = false;

        // Status var that determines whether the app has connected to all requisite services
        this.ready = false;

        // Bucket for prerequisite service connections that block app.ready
        this._serviceConnectors = [];
    }

    /**
     * Applies environment overrides to the app configuration
     * @param env
     * @param config
     */
    _applyEnvConfig(env, config) {
        if (config[env]) {
            if (Cluster.isMaster) {
                this.log('Using environment configuration:', env); // use STDERR so reports and stuff don't get noisy
            }
            this.copy(config, config[env]);
            this.currentEnvironment = env;
        } else {
            throw new Error('Oops! Unknown environment given (not in config): '+env);
        }
    }

    /**
     * Connect to all of the dependant services (e.g. mongo, rabbit, redis, etc)
     * @param [callback] - Will always fire, even if already connected
     * @returns Promise
     */
    connectToServices(callback) {
        return new Promise((resolve, reject) => {
            // Check the dependant service states
            if (this.ready) {
                // App is ready, just callback cuz we're good
                if (callback) return callback();
                return resolve();
            } else {

                // Register the callback when the app finishes initializing
                const ready = () => {
                    if (callback) return callback();
                    return resolve();
                };
                this.once('ready', ready);

                // Check the connection progress
                if (!this._connecting) {
                    this._connecting = true;

                    Promise
                        .all(this._serviceConnectors.map(fn => {
                            // 99% of the time, AsyncFunctions and Promises are interchangeable
                            // but not in Promise.all !
                            if (fn[Symbol.toStringTag] === "AsyncFunction") {
                                return new Promise((resolve, reject) => {
                                    fn().then(resolve, reject);
                                });
                            } else if (fn[Symbol.toStringTag] === "Promise") {
                                return fn;
                            } else {
                                this.log('[OkanjoApp] Callbacks deprecated: Service connectors should return async functions or promises');
                                return Util.promisify(fn);
                            }
                        }))
                        .then(() => {
                            this.ready = true;
                            this._connecting = false;
                            this.emit('ready');
                        }, (err) => {
                            this._connecting = false;
                            this.removeListener('ready', ready);
                            this.emit('error', err);
                            if (callback) return callback(err);
                            return reject(err);
                        })
                    ;
                }
            }
        });
    }

    /**
     * Registers a service connector that will be called when connectToServices is called
     * @param connector – Fired when app should connect to services
     */
    registerServiceConnector(connector) {
        this._serviceConnectors.push(connector);
        // TODO - consider firing the connector is the application has already started (e.g. late connector)
    }

    //endregion

    //region Utility Functions

    /**
     * Simple, deep, key-value copier
     * @param {*} destination – Target object or empty to make brand new copy
     * @param {*} source – Object to make a duplicate of
     * @return {*} – The resulting object, which might be the same as dest unless source was a value not a reference
     * @author Kevin Fitzgerald
     */
    static copy(destination, source) {
        if (source !== null && typeof source === "object") {
            if (Array.isArray(source)) {
                destination = destination || [];
                source.forEach((val, index) => {
                    destination[index] = OkanjoApp.copy(destination[index], val);
                });
            } else {
                destination = destination || {};
                Object.keys(source).forEach((key) => {
                    destination[key] = OkanjoApp.copy(destination[key], source[key]);
                });
            }
        } else {
            destination = source;
        }

        return destination;
    }

    /**
     * Flattens a multi-dimensional object into a single list of key-value pairs (e.g. meta.boobs => meta_boobs)
     * @param input
     * @param options
     * @return {*}
     */
    static flattenData(input, options = {}) {
        const output = {};

        if (input !== undefined && input !== null) {
            Object.keys(input).forEach((key) => {
                // Convert object ids to hex strings
                if (input[key] instanceof Date) {
                    if (options.dateToIso) {
                        output[key] = input[key].toISOString(); // convert to iso
                    } else {
                        output[key] = input[key]; // as-is
                    }
                } else if (typeof input[key] === "object" && input[key] && input[key].constructor && input[key].constructor.name === "ObjectID") { // Object.create(null) is truthy, but .constructor is undefined
                    output[key] = input[key].toString();
                } else if (typeof input[key] === 'object') {
                    // Make child objects flat too (always returns object so Object.keys is safe)
                    const childObject = OkanjoApp.flattenData(input[key], options);
                    Object.keys(childObject).forEach((childKey) => {
                        output[key + '_' + childKey] = childObject[childKey];
                    });
                } else {
                    // Copy value references
                    output[key] = input[key];
                }
            });
        }

        return output;
    }

    //endregion

    //region Reporting Functions

    /**
     * Reports whatever to Sentry
     * @example `app.report("message", err, { data })`
     */
    async report() {

        const agg = { err: undefined, meta: { }, reported: false };

        // Capture where we reported from
        const reportStack = {};
        Error.captureStackTrace(reportStack, this.report);
        reportStack.name = "Reported By:";

        // Show to the console unless ENV var is set to supress
        if (!process.env.SILENCE_REPORTS) {
            console.error('');
            console.error('/------------------------------[ REPORT ]--------------------------------\\');

            // Pick the error out and add the rest to the meta array
            for (let i = 0; i < arguments.length; i++) {

                if (typeof arguments[i] === "object" && arguments[i] instanceof Error) {
                    console.error(arguments[i].stack);
                    agg.err = arguments[i];
                } else {
                    console.error(Util.inspect(arguments[i], {colors: true, depth: 5}));
                    agg.meta['arg' + i] = arguments[i];
                }
            }

            console.error('');
            console.error(reportStack.stack);

            console.error('\\------------------------------------------------------------------------/');
            console.error('');
        }

        // Report as exception so we always know who was responsible for this
        if (this._reportToSentry) {

            // Always get a stack trace
            if (agg.err === undefined) {
                let derivedName = Object.values(agg.meta).find(v => typeof v === "string" && v.length > 0) || "???";
                agg.err = new Error('Report: ' + derivedName);
            }

            // Merge global context
            Object.keys(this.reportingContext).forEach((i) => {
                agg.meta[i] = this.reportingContext[i];
            });

            // Stick in a report stack just to make it easier to figure out how we got to this report
            agg.meta.report_stack = reportStack.stack;

            const data = {
                // user
                // req
                // tags
                // fingerprint
                // level
                extra: agg.meta
            };

            try {
                const eventId = await this._captureException(agg.err, data);
                console.error(' >> Reported as ', eventId);
                agg.eventId = eventId;
                agg.reported = true;
            } catch(err) /* istanbul ignore next: out of scope */ {
                console.error(' >> Failed to report to sentry!', err instanceof Error ? err.stack : err);
            }
        }

        return agg;
    }

    //noinspection JSMethodCanBeStatic
    /**
     * Inspects whatever you give it. Formerly app.inspect(...)
     */
    dump(...args) {
        for (let i = 0; i < args.length; i++) {
            console.error(typeof args[i] === "object" && args[i] instanceof Error ? args[i].stack : Util.inspect(args[i], { colors: true, depth: 5, customInspect: false }) );
        }
    }

    //noinspection JSMethodCanBeStatic
    /**
     * Wrapper around console.error for application logging, so they can be silenced for tests and stuff
     * @param args
     */
    log(...args) {
        if (!process.env.SILENCE_REPORTS) {
            console.error.apply(console, args);
        }
    }

    /**
     * Will change Sentry reporting status. Will bind process unhandled exceptions if enabled.
     * @param enabled
     */
    updateReportingStatus(enabled) {
        this._reportToSentry = enabled || false;

        if (this._reportToSentry) {
            this.log(` > ${Cluster.isWorker ? (process.env.worker_type || 'worker') : 'main' }: Will report uncaught exceptions, starting now.`);
            /* istanbul ignore next: actually triggering an uncaught exception is impossible with mocha */
            process.once('uncaughtException', (err) => this._reportUncaughtException(err));
        }
    }

    /**
     * Report uncaught exceptions and die when one happens
     * @param err
     */
    async _reportUncaughtException(err) {
        // Tell us everything we are doing wrong
        let exitAfterReport = arguments.length <= 1;

        console.error('');
        console.error('/------------------------------[ FATAL EXCEPTION ]--------------------------------\\');
        console.error(err.stack);
        console.error('\\---------------------------------------------------------------------------------/');
        console.error('');

        const eventId = await this._captureException(err, { extra: this.reportingContext });
        console.error(' >> Reported uncaught exception as ', eventId);
        if (exitAfterReport) process.exit(1);
    }

    /**
     * Merges the given context information into the reporting context data
     * @param context
     */
    setReportingContext(context) {
        Object.keys(context).forEach(function(i) {
            this.reportingContext[i] = context[i];
        }, this);
    }

    //endregion
}

/**
 * Response generator
 * @type {Boom}
 * @see https://github.com/hapijs/boom
 */
OkanjoApp.response = require('@hapi/boom');

/**
 * Returns a consistent response payload, like Booms but not
 * @param data
 * @param statusCode
 * @param h – Response toolkit
 * @returns {*}
 */
OkanjoApp.response.custom = (data, statusCode, h) => {
    const output = { statusCode, error: null, data };
    if (h) return h.response(output).code(statusCode);
    return output;
};

/**
 * Creates a 200-ok response object
 * @param {*} data - Response data
 * @param {*} h – Response toolkit
 * @return {{statusCode: number, error: null, data: *}}
 */
OkanjoApp.response.ok = (data, h) => {
    return OkanjoApp.response.custom(data, 200, h);
};

/**
 * Creates a 201-created response object
 * @param {*} data - Response data
 * @param {*} h – Response toolkit
 * @return {{statusCode: number, error: null, data: *}}
 */
OkanjoApp.response.created = (data, h) => {
    return OkanjoApp.response.custom(data, 201, h);
};

/**
 * Helper to format an object or an array of objects using a closure function
 * @param {Object|Object[]} obj - Object to format
 * @param {function(obj:Object)} closure - Called on each object, expects a return value of the formatted object
 * @return {*}
 */
OkanjoApp.response.formatForResponse = (obj, closure) => {
    let out;
    if (Array.isArray(obj)) {
        // ARRAY (recursive)
        out = [];
        for(let i = 0; i < obj.length; i++) {
            out.push(OkanjoApp.response.formatForResponse(obj[i], closure));
        }
    } else {
        // SINGLE
        if (!obj) {
            out = null;
        } else {
            // Object-specific inclusion function
            out = closure(obj);

            // Automatically add auditing fields
            if (out && obj.created !== undefined) { out.created = obj.created; }
            if (out && obj.updated !== undefined) { out.updated = obj.updated; }
        }
    }
    return out;
};

module.exports = OkanjoApp;